File: ios/Classes/rasp/NetworkThreatDetector.swift
import Foundation
import Darwin

/// Detects network-level threats used during reverse engineering from desktop.
public class NetworkThreatDetector {

    private static let _k: [Int] = [0x32 + 0x1C, 0x41 + 0x12, 0x24 + 0x24, 0x3E + 0x0E, 0x22 + 0x22]
    private static func d(_ e: [Int]) -> String {
        String(e.enumerated().map { i, v in Character(UnicodeScalar(v ^ _k[i % _k.count])!) })
    }

    // VPN interface prefixes (encoded)
    private static let vpnPrefixes: [String] = [
        d([59,39,61,34]),    // utun
        d([62,35,56]),       // ppp
        d([39,35,59,41,39]), // ipsec
        d([58,50,56]),       // tap
        d([58,38,38]),       // tun
    ]

    public static func check() -> [String: Any] {
        let proxyDetected = checkProxy()
        let vpnDetected = checkVpn()

        return [
            "proxyDetected": proxyDetected,
            "vpnDetected": vpnDetected,
            "detected": proxyDetected || vpnDetected
        ]
    }

    public static func checkSimple() -> Bool {
        return checkProxy() || checkVpn()
    }

    private static func checkProxy() -> Bool {
        guard let proxySettings = CFNetworkCopySystemProxySettings()?.takeRetainedValue() as? [String: Any] else {
            return false
        }

        if let httpProxy = proxySettings[kCFNetworkProxiesHTTPProxy as String] as? String,
           !httpProxy.isEmpty {
            if let httpEnabled = proxySettings[kCFNetworkProxiesHTTPEnable as String] as? Int,
               httpEnabled == 1 {
                return true
            }
        }

        if let httpsProxy = proxySettings[kCFNetworkProxiesHTTPSProxy as String] as? String,
           !httpsProxy.isEmpty {
            if let httpsEnabled = proxySettings[kCFNetworkProxiesHTTPSEnable as String] as? Int,
               httpsEnabled == 1 {
                return true
            }
        }

        if let socksProxy = proxySettings[kCFNetworkProxiesSOCKSProxy as String] as? String,
           !socksProxy.isEmpty {
            if let socksEnabled = proxySettings[kCFNetworkProxiesSOCKSEnable as String] as? Int,
               socksEnabled == 1 {
                return true
            }
        }

        return false
    }

    private static func checkVpn() -> Bool {
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let firstAddr = ifaddr else {
            return false
        }
        defer { freeifaddrs(ifaddr) }

        var addr = firstAddr
        while true {
            let name = String(cString: addr.pointee.ifa_name)
            let flags = Int32(addr.pointee.ifa_flags)
            let isUp = (flags & IFF_UP) != 0

            if isUp {
                for prefix in vpnPrefixes {
                    if name.hasPrefix(prefix) {
                        return true
                    }
                }
            }

            guard let next = addr.pointee.ifa_next else { break }
            addr = next
        }

        return false
    }
}

File: linux/rasp/network_threat_detector.cc
#include "network_threat_detector.h"

#include <cstdlib>
#include <cstring>
#include <fstream>
#include <string>
#include <ifaddrs.h>
#include <net/if.h>
#include "../shield_codec.h"

namespace flutter_neo_shield {

// Encoded proxy environment variable names
static const std::string kProxyVars[] = {
  ShieldCodec::Decode({38,39,60,60,27,62,33,39,52,61}),      // http_proxy
  ShieldCodec::Decode({38,39,60,60,55,17,35,58,35,60,55}),   // https_proxy
  ShieldCodec::Decode({6,7,28,28,27,30,1,7,20,29}),          // HTTP_PROXY
  ShieldCodec::Decode({6,7,28,28,23,17,3,26,3,28,23}),       // HTTPS_PROXY
  ShieldCodec::Decode({15,31,4,19,20,28,28,16,21}),          // ALL_PROXY
  ShieldCodec::Decode({47,63,36,19,52,60,60,48,53}),         // all_proxy
};
static const size_t kProxyVarsCount = sizeof(kProxyVars) / sizeof(kProxyVars[0]);

// Encoded VPN interface prefixes
static const std::string kVpnPrefixes[] = {
  ShieldCodec::Decode({58,38,38}),  // tun
  ShieldCodec::Decode({58,50,56}),  // tap
  ShieldCodec::Decode({62,35,56}),  // ppp
  ShieldCodec::Decode({57,52}),     // wg
};
static const size_t kVpnPrefixesCount = sizeof(kVpnPrefixes) / sizeof(kVpnPrefixes[0]);

bool NetworkThreatDetector::CheckSimple() {
  return CheckProxy() || CheckVpn();
}

/// Check for proxy configuration via environment variables.
bool NetworkThreatDetector::CheckProxy() {
  for (size_t i = 0; i < kProxyVarsCount; i++) {
    const char* val = getenv(kProxyVars[i].c_str());
    if (val && strlen(val) > 0) {
      return true;
    }
  }

  return false;
}

/// Check for VPN interfaces via getifaddrs.
bool NetworkThreatDetector::CheckVpn() {
  struct ifaddrs* ifaddr;
  if (getifaddrs(&ifaddr) != 0) return false;

  bool found = false;

  for (struct ifaddrs* ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
    if (!ifa->ifa_name) continue;

    if (ifa->ifa_flags & IFF_UP) {
      std::string name(ifa->ifa_name);
      for (size_t i = 0; i < kVpnPrefixesCount; i++) {
        if (name.compare(0, kVpnPrefixes[i].size(), kVpnPrefixes[i]) == 0) {
          found = true;
          break;
        }
      }
      if (found) break;
    }
  }

  freeifaddrs(ifaddr);
  return found;
}

}  // namespace flutter_neo_shield
